WooCommerce/Classes/Tools/InfiniteScroll/PaginationTracker.swift

```diff
@@ -0,0 +1,152 @@
+import Yosemite
+
+/// Delegate of `PaginationTracker` that implements syncing per page number and size.
+protocol PaginationTrackerDelegate: class {
+    typealias SyncCompletion = (Result<Bool, Error>) -> Void
+
+    /// Syncs a page by the number and size, and returns an error or whether there might be the next page on completion.
+    ///
+    /// - Parameter reason: A value passed from `resync` or `syncFirstPage`. This can
+    ///                     be used to decide how to perform the sync.
+    /// - Parameter onCompletion: called on sync completion that returns either an error or a boolean that indicates
+    ///                           whether there might be the next page to sync.
+    func sync(pageNumber: Int, pageSize: Int, reason: String?, onCompletion: SyncCompletion?)
+}
+
+/// Keeps track of the pagination for API syncing to support infinite scroll.
+final class PaginationTracker {
+    /// Default Settings
+    private enum Defaults {
+        static let pageFirstIndex = Store.Default.firstPageNumber
+        static let pageSize = 25
+    }
+
+    /// First Page Index
+    private let pageFirstIndex: Int
+
+    /// Number of elements retrieved per request.
+    private let pageSize: Int
+
+    /// Syncing delegate
+    weak var delegate: PaginationTrackerDelegate?
+
+    /// Maps from a page number to the last synced date.
+    private var syncedDatesByPage = [Int: Date]()
```

Comment:
I don't see the `Date` value being used anywhere. Is this a placeholder for this?

> I don't feel like we need to sync an earlier page when scrolling to it, because if an earlier page has changes, we probably want to fetch the other earlier pages anyway.

There's also a chance I misunderstood what you meant there ☝️ 😅

Reply:
> There's also a chance I misunderstood what you meant there ☝️ 😅

The type should just be `IndexSet()` similar to `pagesBeingSynced` here, thanks for pointing it out! I forgot to update after borrowing from `SyncingCoordinator`, changes made in https://github.com/woocommerce/woocommerce-ios/pull/2750/commits/85d2cbc94229ee56a5487eb03b9996b1dfb943f8. I think we're on the same page that the caching behavior (thus recording the last sync date) isn't necessary. Sorry for the long and not so clear sentence before!